### PR TITLE
Localization: add ndt map files for pyramid_map module

### PR DESCRIPTION
**Pull request by tthhee**

### CODE REVIEW

File: modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix.h [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+  size_t GetBinarySize() const;
+  /**@brief Overloading the assign operator. */
+  NdtMapSingleCell& operator=(const NdtMapSingleCell& ref);
+
+  /**@brief Combine two NdtMapSingleCell instances (Reduce). */
+  static void Reduce(NdtMapSingleCell* cell, const NdtMapSingleCell& cell_new);
+
+  /**@brief Add an sample to the single 3d map cell. */
+  inline void AddSample(const float intensity, const float altitude,
+                        const Eigen::Vector3f centroid, bool is_road = false);
```
Comment (anyong298) [inline modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix.h]: Eigen::Vector3f pass by reference

Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix.h]: fixed

File: modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix.h [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  road_pt_count_ += road_pt_count;
+}
+
+inline void NdtMapSingleCell::MergeCell(const NdtMapSingleCell& cell_new) {
+  MergeCell(cell_new.intensity_, cell_new.intensity_var_,
+            cell_new.road_pt_count_, cell_new.count_, cell_new.centroid_,
+            cell_new.centroid_average_cov_);
+}
+
+inline void NdtMapSingleCell::CentroidEigenSolver(
```
Comment (jinghaomiao) [inline modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix.h]: Avoid inline functions with complicated logic. This applies to all where applicable.

File: modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix.cc [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+}
+
+void NdtMapMatrix::Reset() { Reset(rows_, cols_); }
+
+void NdtMapMatrix::Init(unsigned int rows, unsigned int cols) {
+  if (map3d_cells_) {
+    delete[] map3d_cells_;
+  }
+  map3d_cells_ = nullptr;
+  map3d_cells_ = new NdtMapCells[rows * cols];
```
Comment (jinghaomiao) [inline modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix.cc]: Use safe pointers

File: modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix_handler.cc [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+namespace apollo {
+namespace localization {
+namespace msf {
+// =================PyramidMapMatrixHandlerSelector=================
+NdtMapMatrixHandlerSelector::NdtMapMatrixHandlerSelector() {}
+
+NdtMapMatrixHandlerSelector::~NdtMapMatrixHandlerSelector() {}
+
+BaseMapMatrixHandler* NdtMapMatrixHandlerSelector::AllocNdtMapMatrixHandler() {
+  return new NdtMapMatrixHandler();
```
Comment (jinghaomiao) [inline modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix_handler.cc]: Use safe pointers

File: modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_pool.cc [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+namespace msf {
+
+/**@brief Constructor
+ * @param <pool_size> The memory pool size.
+ * @param <thread_size> The thread pool size.
+ */
+NdtMapNodePool::NdtMapNodePool(unsigned int pool_size, unsigned int thread_size)
+    : BaseMapNodePool(pool_size, thread_size) {}
+
+BaseMapNode* NdtMapNodePool::AllocNewMapNode() { return new NdtMapNode(); }
```
Comment (jinghaomiao) [inline modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_pool.cc]: Use safe pointers.

**Comment by jinghaomiao** [comment]
@tthhee Please check all your pull requests whether merged or not and use safe pointers for all "new" pointers. Thanks!

File: modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix_handler.cc (hunk 4 above)
Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix_handler.cc]: The caller will use safe pointers to pack it.

File: modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_pool.cc (hunk 5 above)
Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_pool.cc]: we will manager the resource in MapNodePool.

File: modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix.cc (hunk 3 above)
Comment (tthhee) [reply modules/localization/msf/local_pyramid_map/ndt_map/ndt_map_matrix.cc]: fixed

**Pull request merged**